app/src/main/java/com/tari/android/wallet/ffi/FFIWallet.kt

```diff
@@ -549,32 +547,31 @@ internal class FFIWallet(
             status
         )
 
-        listenerAdapter?.onTxFinalized(completed)
-        tx.destroy()
+        listenerAdapter?.run { Handler(Looper.getMainLooper()).post { onTxFinalized(completed) } }
     }
 
     fun onDirectSendResult(bytes: ByteArray, success: Boolean) {
         Logger.i("Direct send result received. Success: $success")
         val txId = BigInteger(1, bytes)
-        listenerAdapter?.onDirectSendResult(txId, success)
+        listenerAdapter?.run { Handler(Looper.getMainLooper()).post { onDirectSendResult(txId, success) } }
     }
 
     fun onStoreAndForwardSendResult(bytes: ByteArray, success: Boolean) {
         Logger.i("Store and forward send result received. Success: $success")
         val txId = BigInteger(1, bytes)
-        listenerAdapter?.onStoreAndForwardSendResult(txId, success)
+        listenerAdapter?.run { Handler(Looper.getMainLooper()).post { onStoreAndForwardSendResult(txId, success) } }
     }
 
     fun onTxCancellation(bytes: ByteArray) {
         Logger.i("Transaction cancelled.")
         val txId = BigInteger(1, bytes)
-        listenerAdapter?.onTxCancellation(txId)
+        listenerAdapter?.run { Handler(Looper.getMainLooper()).post { onTxCancellation(txId) } }
     }
 
     fun onBaseNodeSyncComplete(bytes: ByteArray, success: Boolean) {
         Logger.i("Base node sync complete. Success: $success")
         val requestId = BigInteger(1, bytes)
-        listenerAdapter?.onBaseNodeSyncComplete(requestId, success)
+        listenerAdapter?.run { Handler(Looper.getMainLooper()).post { onBaseNodeSyncComplete(requestId, success) } }
```

Comment:
@kukabi I might be thinking that it is possible to wrap all the adapter invocations into a Java's proxy ( https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Proxy.html ) that dispatches the given calls to the main thread, thus eliminating the "this method should be wrapped too" problem altogether as well as eliminating risks when new methods are being added to the wallet interface, what do you think?

Reply:
@nyarian Aren't proxies good when you don't have access to the implementation? We can fix the actual code here. We can find a way of enforcing the thread creation but I think what we have is just good enough at the moment.

Reply:
It's just already third PR after the initial one which adds a Handler delegation to several others adapter invocations, so it directs to some thoughts

Proxy is proposed here because it is a handy way not to override methods by hand. I've seen such practice incorporated to wrap a `Presenter` interface inheritor into a dynamic proxy instance that delegates all the calls from the UI component (and, thus, from the UI thread) to the background one, and the reverse proxy wraps the `View` interface inheritor. So, basically, a single proxy implementation, which is simply instantiated for each interface you need

But "good enough" is "good enough" then :)